## Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

![https://s3-us-west-2.amazonaws.com/secure.notion-static.com/069818d1-0e5c-4d87-a461-0de584ab9c33/Untitled_(33).png](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/069818d1-0e5c-4d87-a461-0de584ab9c33/Untitled_(33).png)

## Задачи

- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц;
- Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.


### Доступ к базе данных

Следуйте инструкции, приложенной ниже

[Подключение к базе данных .pdf](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/d66eb372-79d0-4799-a527-6ebfe46cbddc/____.pdf)

## Цель исследования

Цель исследования - проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Исследование таблиц


In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Таблица books

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

In [3]:
query = '''
SELECT 
    *
FROM books
'''

In [4]:
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

В базе содержится информация о 1000 книг.

### Таблица authors

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

In [5]:
query = '''
SELECT 
    author_id,
    author
FROM authors
'''

In [6]:
pd.io.sql.read_sql(query, con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

В базе содержится информация о 636 авторах.

### Таблица publishers

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;


In [7]:
query = '''
SELECT 
    *
FROM publishers
'''

In [8]:
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

В базе содержится информация о 340 издательствах.

### Таблица ratings

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

In [9]:
query = '''
SELECT 
    *
FROM ratings
'''

In [10]:
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

В базе содержится информация о 6456 оценках книг.

### Таблица reviews

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [11]:
query = '''
SELECT 
    *
FROM reviews
'''

In [12]:
pd.io.sql.read_sql(query, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

В базе содержится информация о 2793 обзорах на книги.

## Расчет количества книг, выпущенных после 1 января 2000 года

In [13]:
query = '''
SELECT 
    COUNT(*) AS cnt
FROM 
    books
WHERE
    publication_date > '2000-01-01'
'''

In [14]:
pd.io.sql.read_sql(query, con = engine)

cnt
0  819

### Вывод

В базе данных содержится информация о 819 книгах, дата публикации которых позже 1 января 2000 года.

In [15]:
# код ревьюера (уникальное количество книг по id):
pd.io.sql.read_sql('''
SELECT 
    COUNT(DISTINCT book_id) AS cnt
FROM 
    books
WHERE
    publication_date > '2000-01-01'
''', con = engine)

cnt
0  819

In [16]:
# код ревьюера (уникальное количество книг по заголовку):
pd.io.sql.read_sql('''
SELECT 
    COUNT(DISTINCT title) AS cnt
FROM 
    books
WHERE
    publication_date > '2000-01-01'
''', con = engine)

cnt
0  818

## Расчет количества обзоров и средней оценки для каждой книги 

In [17]:
query = '''
SELECT
    books.title AS book_name,
    subquery_reviews.reviews_cnt AS reviews_cnt,
    AVG(ratings.rating) AS avg_rating
FROM 
    books
    LEFT JOIN (
        SELECT
            book_id,
            COUNT(review_id) AS reviews_cnt
        FROM
            reviews
        GROUP BY
            book_id
        ) AS subquery_reviews ON subquery_reviews.book_id = books.book_id
    LEFT JOIN ratings ON ratings.book_id = books.book_id
WHERE reviews_cnt IS NOT NULL
GROUP BY
    books.title,
    subquery_reviews.reviews_cnt
ORDER BY
    reviews_cnt DESC
'''

In [18]:
pd.io.sql.read_sql(query, con = engine)

book_name  reviews_cnt  \
0                              Twilight (Twilight  #1)            7   
1                                      Eat  Pray  Love            6   
2    Harry Potter and the Chamber of Secrets (Harry...            6   
3    Harry Potter and the Prisoner of Azkaban (Harr...            6   
4                            Outlander (Outlander  #1)            6   
..                                                 ...          ...   
989                                The Social Contract            1   
990  To Green Angel Tower (Memory  Sorrow  and Thor...            1   
991               Tsubasa: RESERVoir CHRoNiCLE  Vol. 1            1   
992                              Wicked: The Grimmerie            1   
993                  Winter Prey (Lucas Davenport  #5)            1   

     avg_rating  
0      3.662500  
1      3.395833  
2      4.287500  
3      4.414634  
4      4.125000  
..          ...  
989    3.500000  
990    4.500000  
991    2.500000  
992    3.500000  
993    4.500000  

[994 rows x 3 columns]

In [19]:
query = '''
SELECT
    books.title AS book_name,
    subquery_reviews.reviews_cnt AS reviews_cnt,
    AVG(ratings.rating) AS avg_rating
FROM 
    books
    LEFT JOIN (
        SELECT
            book_id,
            COUNT(review_id) AS reviews_cnt
        FROM
            reviews
        GROUP BY
            book_id
        ) AS subquery_reviews ON subquery_reviews.book_id = books.book_id
    LEFT JOIN ratings ON ratings.book_id = books.book_id
GROUP BY
    books.title,
    subquery_reviews.reviews_cnt
ORDER BY
    avg_rating DESC
'''
pd.io.sql.read_sql(query, con = engine)

book_name  reviews_cnt  \
0                                  Homage to Catalonia          2.0   
1                                                March          2.0   
2    In the Hand of the Goddess (Song of the Liones...          2.0   
3                 Welcome to Temptation (Dempseys  #1)          2.0   
4    Stone of Farewell (Memory  Sorrow  and Thorn  #2)          1.0   
..                                                 ...          ...   
995  The World Is Flat: A Brief History of the Twen...          3.0   
996                                              Junky          2.0   
997                  His Excellency: George Washington          2.0   
998                                      Drowning Ruth          3.0   
999                               Harvesting the Heart          2.0   

     avg_rating  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
995        2.25  
996        2.00  
997        2.00  
998        2.00  
999        1.50  

[1000 rows x 3 columns]

### Вывод

При выводе отсортированной таблицы можно увидеть, что максимальное количество обзоров не слишком велико - максимальное число обзоров (7) пришлось на книгу Сумерки. Также есть 6 книг совсем без обзоров.

## Определение издательства, которое выпустило наибольшее число книг толще 50 страниц

In [20]:
query = '''
SELECT
    publishers.publisher,
    COUNT(book_id) AS books_cnt
FROM
    books
    LEFT JOIN publishers ON publishers.publisher_id = books.publisher_id
WHERE 
    num_pages > 50
GROUP BY
    publishers.publisher
ORDER BY
    books_cnt DESC
LIMIT 1
'''

In [21]:
pd.io.sql.read_sql(query, con = engine)

publisher  books_cnt
0  Penguin Books         42

### Вывод

Издательство, которое выпустило наибольшее число книг толще 50 страниц - это Penguin Books. Оно выпустило 42 книги. Это достаточно старое и серьезное издательство, поэтому неудивительно, что оно на первом месте.

## Определение автора с самой высокой средней оценкой книг (с учетом только книг с 50 и более оценками)

In [22]:
query = '''
SELECT
    authors.author,
    AVG(subquery_ratings.avg_rating) AS avg_rating
FROM
    books
    LEFT JOIN (
        SELECT 
            book_id,
            AVG(rating) AS avg_rating,
            COUNT(rating_id) AS rating_cnt
        FROM
            ratings
        GROUP BY
            book_id
    ) AS subquery_ratings ON subquery_ratings.book_id = books.book_id
    LEFT JOIN authors ON authors.author_id = books.author_id
WHERE
    rating_cnt >= 50
GROUP BY
    author
ORDER BY
    avg_rating DESC
LIMIT 1
'''

In [23]:
pd.io.sql.read_sql(query, con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.283844

### Вывод

Самая высокая средняя оценка книг у автора Д. Роулинг. Это довольно ожидаемый результат - серия книг о Гарри Поттере действительно пользуется популярностью по всему миру. Средняя оценка ее книг составляет 4.28 балла.

## Расчет среднего количества обзоров от пользователей, которые поставили больше 50 оценок

In [24]:
query = '''
WITH
    ratings_subquery AS (
        SELECT
            ratings.username,
            COUNT(rating_id) AS ratings_cnt
        FROM
            ratings
        GROUP BY
            ratings.username
        HAVING
            COUNT(rating_id) > 50
        ),
    reviews_subquery AS (
        SELECT
            reviews.username,
            COUNT(review_id) AS reviews_cnt
        FROM
            reviews
        GROUP BY
            reviews.username
    )
SELECT
    AVG(reviews_cnt) as avg_reviews_cnt
FROM
    ratings_subquery
    LEFT JOIN reviews_subquery ON reviews_subquery.username = ratings_subquery.username
'''

In [25]:
pd.io.sql.read_sql(query, con = engine)

avg_reviews_cnt
0        24.333333

### Вывод

Найдена информация об активных пользователях, которые поставили более 50 оценок. 
В среднем такие пользователи оставляют 24 обзора.

## Общий вывод

В ходе данного проекта была проанализирована база данных крупного сервиса для чтения книг по подписке. В ней содержится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. В резулате анализа получены следующие результаты:

1. В БД нового сервиса содержится информация о 1000 книг. Для крупного сервиса это не так много. 
2. В БД содержится информация о 636 авторах,  340 издательствах,  6456 оценках книг и о 2793 обзорах на книги.
3. Порядка 82% книг выпущены за последние 20 лет (819).
4. Для большинства книг в БД есть оценки и обзоры (99%).
5. Максимальное количество обзоров не слишком велико - 7, оно пришлось на книгу Сумерки. 
6. Издательство, которое выпустило наибольшее число книг толще 50 страниц - это Penguin Books. Оно выпустило 42 книги, что составляет 4% от выборки.
7. Самые высокие оценки у автора Д. Роулинг. Средняя оценка ее книг составляет 4.28 балла.
8. В базе есть информация об активных пользователях - купленный сервис достаточно "живой", но располагает не слишком большой подборкой книг.